client/src/containers/AuthContainer.test.tsx

```diff
@@ -98,3 +88,122 @@ describe("verifying valid input for signup page", () => {
     });
   });
 });
+
+describe("signUp component", () => {
+  describe("handleSubmit function", () => {
+    it("should show password mismatch alert", async () => {
+      screen.getByTestId("SignUpTabButton").click();
+
+      const fields = [
+        ["passwordForm", "password"],
+        ["confirmPassForm", "passwordd"],
+      ];
+      fields.forEach((pair) => {
+        fireEvent.change(
+          screen
+            .getByTestId(pair[0])
+            .querySelector("input") as HTMLInputElement,
+          { target: { value: pair[1] } }
+        );
+      });
+
+      const form = screen.getByTestId("form");
+      const mockAlert = jest
+        .spyOn(window, "alert")
+        .mockImplementation(() => {});
+
+      fireEvent.submit(form);
+      expect(mockAlert).toBeCalledWith("Passwords do not match.");
+    });
+
+    it("should show backend error alert on backend error", async () => {
+      screen.getByTestId("SignUpTabButton").click();
+
+      const form = screen.getByTestId("form");
+      const msg = "test";
+      const mockSignUp = jest
+        .spyOn(ServerApi.prototype, "signUp")
+        .mockImplementation(() =>
+          Promise.resolve({
+            error: { response: { data: { message: msg } } } as AxiosError,
+          })
+        );
+      const mockAlert = jest
+        .spyOn(window, "alert")
+        .mockImplementation(() => {});
+
+      fireEvent.submit(form);
+      /**
+       * Since fireEvent.submit is not async function we cannot use await
+       * Without an await, expect(mockAlert) assertion can run before
+       * fireEvent.submit can complete execution causing the test to fail.
+       * Therefore, we await on dummy Promise to delay execution while
+       * fireEvent.submit completes.
+       */
+      await new Promise((r) => setTimeout(r, 1));
+      expect(mockSignUp).toBeCalled();
+      expect(mockAlert).toBeCalledWith(`Message: ${msg}`);
+    });
+
+    it("should show email confirmation alert on success", async () => {
+      screen.getByTestId("SignUpTabButton").click();
+
+      const form = screen.getByTestId("form");
+      const mockSignUp = jest
+        .spyOn(ServerApi.prototype, "signUp")
+        .mockImplementation(() =>
+          Promise.resolve({ response: {} as AxiosResponse })
+        );
+      const mockAlert = jest
+        .spyOn(window, "alert")
+        .mockImplementation(() => {});
+
+      fireEvent.submit(form);
+      await new Promise((r) => setTimeout(r, 1));
+      expect(mockSignUp).toBeCalled();
+      expect(mockAlert).toBeCalledWith("Email confirmation sent!");
```

Comment:
If you assert that mockSignup was called with certain params (the expected ones) you can avoid mockAlert and asserting that it was triggered. Testing the call params is more robust.

Reply:
How would we test the different params since we can't actually test what happens to mockSignUp based on these different inputs?

Also, I'm not sure how to differentiate a successful sign up from a failed signup without checking the alert messages logged.

Reply:
`expect(mockSignUp).toBeCalledWith(expectedEmail, expectedUserName, ...)` where expectedEmail is the email that the form input was set to. So if in your test, you set the users email in the form to be 'foo@mail.utoronto.ca`, the mock signup function should be called with  'foo@mail.utoronto.ca` as the first param (similar logic extends to username, password, passwordConfirmation, first and last name). If the mock signup function has not been called, that means that some component level validation prevented that from happening (e.g. the email format was bad). 

So if the mock signup function was called, that means that it was a successful signup. If there was not successful signup, e.g. email format was invalid, or the password was too short, then the mockSignup function should not be called.